source/mesh_refinement/composition_approximate_gradient.cc

```diff
@@ -0,0 +1,157 @@
+/*
+  Copyright (C) 2015 - 2016 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/mesh_refinement/composition_approximate_gradient.h>
+
+#include <deal.II/dofs/dof_tools.h>
+#include <deal.II/numerics/derivative_approximation.h>
+
+namespace aspect
+{
+  namespace MeshRefinement
+  {
+    template <int dim>
+    void
+    CompositionApproximateGradient<dim>::execute(Vector<float> &indicators) const
+    {
+      AssertThrow (this->n_compositional_fields() >= 1,
+                   ExcMessage ("This refinement criterion can not be used when no "
+                               "compositional fields are active!"));
+      indicators = 0;
+      // create a vector with the requisite ghost elements
+      // and use it for estimating the gradients of compositional field
+      LinearAlgebra::BlockVector vec(this->introspection().index_sets.system_partitioning,
+                                     this->introspection().index_sets.system_relevant_partitioning,
+                                     this->get_mpi_communicator());
+      for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+        {
+          const unsigned int block_idx = this->introspection().block_indices.compositional_fields[c];
+          vec.block(block_idx) = this->get_solution().block(block_idx);
+          vec.compress(VectorOperation::insert);
+
+          Vector<float> indicators_tmp(indicators);
+          DerivativeApproximation::approximate_gradient(this->get_mapping(),
+                                                        this->get_dof_handler(),
+                                                        vec,
+                                                        indicators_tmp,
+                                                        this->introspection().component_indices.compositional_fields[c]);
+
+          const double power = 1.0 + dim / 2.0;
+          unsigned int i = 0;
```

Comment:
I this I still need this, but I should move this declaration integer i  later but right before
          for (cell = this->get_dof_handler().begin_active(); cell != endc; ++cell, ++i)
            if (cell->is_locally_owned())
           indicators_tmp(i) *= std::pow(cell->diameter(), power) * composition_scaling_factors[c];